### PR TITLE
feat(xrpl): Wallet implementation

### PR DESCRIPTION
**Pull request by GuillemGarciaDev**
# feat(xrpl): Wallet implementation

This PR aims to implement the `Wallet` API to generate and build wallets from seeds or mnemonic and sign transactions with them. This first implementation includes the following functions:

- `NewWallet`
- `NewWalletFromSeed`
- `NewWalletFromSecret`
- `NewWalletFromMnemonic`
- `Sign`
- `GetAddress`

## Changes

- Replicated Wallet API from `xrpl.js`.
- Added all wallet constructors (without RFC1751 mnemonic)
- Added `Sign` method.

## Notes (optional)

There are other interesting methods that should be implemented such as: `VerifyTransaction` or `GetXAddress`.

### CODE REVIEW

File: xrpl/wallet.go [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+type Wallet struct {
+	PublicKey      string
+	PrivateKey     string
+	ClassicAddress string
+	Seed           string
+}
+
+// Creates a new random Wallet. In order to make this a valid account on ledger, you must
+// Send XRP to it.
+func NewWallet(alg addresscodec.CryptoAlgorithm) (Wallet, error) {
```
Comment (florent-uzio) [inline xrpl/wallet.go:29]: Would it make more sense to call it `Generate` like in xrpl.js ?

I am new to Go but I would expect we would import the `Wallet` and then write something like:

```go
import ( "wallet" )

wallet.Generate()
```

or also (based on your example file):

```go
import ( "xrpl" )

xrpl.Wallet.Generate()
```

File: xrpl/wallet.go [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+	seed, err := keypairs.GenerateSeed("", alg)
+	if err != nil {
+		return Wallet{}, err
+	}
+	return NewWalletFromSeed(seed, "")
+}
+
+// Derives a wallet from a seed.
+// Returns a Wallet object. If an error occurs, it will be returned.
+func NewWalletFromSeed(seed string, masterAddress string) (Wallet, error) {
```
Comment (florent-uzio) [inline xrpl/wallet.go:39]: Would `FromSeed` be enough following my previous comment?

File: xrpl/wallet.go [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+		PrivateKey:     privKey,
+		Seed:           seed,
+		ClassicAddress: classicAddr,
+	}, nil
+
+}
+
+// Derives a wallet from a secret (AKA a seed).
+// Returns a Wallet object. If an error occurs, it will be returned.
+func NewWalletFromSecret(seed string) (Wallet, error) {
```
Comment (florent-uzio) [inline xrpl/wallet.go:67]: FromSecret?

File: xrpl/wallet.go [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+	return NewWalletFromSeed(seed, "")
+}
+
+// Derives a wallet from a bip39 or RFC1751 mnemonic (Defaults to bip39).
+// Returns a Wallet object. If an error occurs, it will be returned.
+func NewWalletFromMnemonic(mnemonic string) (*Wallet, error) {
```
Comment (florent-uzio) [inline xrpl/wallet.go:73]: `FromMnemonic` ?

**Review by florent-uzio — APPROVED** [review]
Very nice PR!
I have some naming comments but no need to block for this.
Approved.

File: xrpl/wallet.go (hunk 1 above)
Comment (GuillemGarciaDev) [reply xrpl/wallet.go:29]: The issue with Go is that exports behave different than in javascript. Every function and struct lives in a package. So, in order to use a function (p.e.) that is stored in "x" package, it needs to be called with:

```go
import (
   "x"
)

func main() {
   x.Method()
}
```

Comment (GuillemGarciaDev) [reply xrpl/wallet.go:29]: we could make a package `wallet` and store everything in there, but calling the methods can be weird.

```go
import (
   "wallet"
)

func main() {
   wallet, err := wallet.GenerateWallet(...)
}
```

**Pull request merged**